ott/core/sinkhorn_lr.py

```diff
@@ -338,15 +426,34 @@ def lr_costs(
       self, ot_prob: linear_problems.LinearProblem, state: LRSinkhornState,
       iteration: int
   ) -> Tuple[jnp.ndarray, jnp.ndarray, jnp.ndarray]:
-    c_q = ot_prob.geom.apply_cost(state.r, axis=1) / state.g[None, :]
-    c_q += (self.epsilon - 1 / self.gamma) * jnp.log(state.q)
-    c_r = ot_prob.geom.apply_cost(state.q) / state.g[None, :]
-    c_r += (self.epsilon - 1 / self.gamma) * jnp.log(state.r)
+    grad_q = ot_prob.geom.apply_cost(state.r, axis=1) / state.g[None, :]
+    grad_q = jnp.where(
+        self.epsilon != 0., grad_q + self.epsilon * jnp.log(state.q), grad_q
+    )
+    if self.gamma_init == "rescale":
+      norm_q = jnp.max(jnp.abs(grad_q)) ** 2
+    grad_r = ot_prob.geom.apply_cost(state.q) / state.g[None, :]
+    grad_r = jnp.where(
+        self.epsilon != 0., grad_r + self.epsilon * jnp.log(state.r), grad_r
+    )
+    if self.gamma_init == "rescale":
+      norm_r = jnp.max(jnp.abs(grad_r)) ** 2
     diag_qcr = jnp.sum(
         state.q * ot_prob.geom.apply_cost(state.r, axis=1), axis=0
     )
+    grad_g = -diag_qcr / state.g ** 2
+    grad_g = jnp.where(
+        self.epsilon != 0., grad_g + self.epsilon * jnp.log(state.g), grad_g
+    )
+    if self.gamma_init == "rescale":
+      norm_g = jnp.max(jnp.abs(grad_g)) ** 2
+    if self.gamma_init == "rescale":
+      self.gamma = self.gamma / max(norm_q, norm_r, norm_g)
     h = diag_qcr / state.g ** 2 - (self.epsilon -
                                    1 / self.gamma) * jnp.log(state.g)
+    c_q = grad_q - (1 / self.gamma) * jnp.log(state.q)
```

Comment:
Would assign `log(state.{q,r,g})` to variables, since they can be reused when computed `grad_{q,r,g}`